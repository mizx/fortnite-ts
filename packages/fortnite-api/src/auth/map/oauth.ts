import { OAuthTokenResponse, OAuthExchangeResponse } from '../response';

export function mapOAuthTokenResponseToAccessToken(response: OAuthTokenResponse) {
  return response.access_token;
}

export function mapOAuthExchangeResponseToExchangeCode(response: OAuthExchangeResponse) {
  return response.code;
}

export function mapOAuthTokenResponseToAuthData(response: OAuthTokenResponse) {
  return {
    accessToken: response.access_token,
    appId: response.in_app_id,
    refreshToken: response.refresh_token,
    accountId: response.account_id,
    expiresAt: new Date(response.expires_at)
  };
}
