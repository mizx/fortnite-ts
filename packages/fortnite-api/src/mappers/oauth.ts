import { OAuthTokenResponse, OAuthExchangeResponse } from '../auth/responses';
import { AuthData, AuthRefreshData } from '../auth/interfaces';

export function mapLauncherTokenToAccessToken(response: OAuthTokenResponse): string {
  return response.access_token;
}

export function mapExchangeTokenToExchangeCode(response: OAuthExchangeResponse): string {
  return response.code;
}

export function mapFortniteTokenToAuthData(response: OAuthTokenResponse): AuthData {
  return {
    accessToken: response.access_token,
    appId: response.in_app_id,
    refreshToken: response.refresh_token,
    accountId: response.account_id,
    expiresAt: new Date(response.expires_at)
  };
}

export function mapRefreshTokenToRefreshData(response: OAuthTokenResponse): AuthRefreshData {
  return {
    accessToken: response.access_token,
    refreshToken: response.refresh_token,
    expiresAt: new Date(response.expires_at)
  };
}
