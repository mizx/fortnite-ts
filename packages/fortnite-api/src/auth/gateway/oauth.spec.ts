import nock from 'nock';
import { OAUTH_TOKEN_ENDPOINT, OAUTH_EXCHANGE_ENDPOINT, getLauncherToken, exchangeToken, getFortniteToken } from './oauth';

const OAUTH_TOKEN_SUCCESS_RESPONSE = `{
  "access_token": "access-token-hash",
  "expires_in": 28800,
  "expires_at": "2018-08-17T14:59:39.368Z",
  "token_type": "bearer",
  "refresh_token": "refresh-token",
  "refresh_expires": 1987200,
  "refresh_expires_at": "2018-09-09T06:59:39.368Z",
  "account_id": "account-id-hash",
  "client_id": "client-id-hash",
  "internal_client": true,
  "client_service": "launcher",
  "lastPasswordValidation": "2018-08-17T06:59:39.368Z",
  "app": "launcher",
  "in_app_id": "app-id-hash"
}`;
const OAUTH_TOKEN_FAILURE_RESPONSE = `{
  "errorCode": "errors.com.epicgames.account.invalid_account_credentials",
  "errorMessage": "Sorry the account credentials you are using are invalid",
  "messageVars": [],
  "numericErrorCode": 18031,
  "originatingService": "com.epicgames.account.public",
  "intent": "prod",
  "error_description": "Sorry the account credentials you are using are invalid",
  "error": "invalid_grant"
}`;
const OAUTH_EXCHANGE_SUCCESS_RESPONSE = `{
  "expiresInSeconds": 299,
  "code": "code",
  "creatingClientId": "client-id"
}`;
const OAUTH_EXCHANGE_FAILURE_RESPONSE = `{
  "errorCode": "errors.com.epicgames.common.oauth.invalid_token",
  "errorMessage": "The OAuthToken you are using is not valid",
  "messageVars": [],
  "numericErrorCode": 1014,
  "originatingService": "com.epicgames.account.public",
  "intent": "prod"
}`;

describe('OAuth Gateway', () => {

  afterEach(() => {
    nock.cleanAll();
  });

  describe('getLauncherToken()', () => {
    const username = 'username';
    const password = 'password';

    beforeEach(() => {
      const url = OAUTH_TOKEN_ENDPOINT;
      interface Body {
        username: string;
        password: string;
      }

      nock(url.origin)
        .persist()
        .post(url.pathname, (body: Body) => body.username === username && body.password === password)
        .reply(200, OAUTH_TOKEN_SUCCESS_RESPONSE)
        .post(url.pathname)
        .reply(400, OAUTH_TOKEN_FAILURE_RESPONSE);
    });

    it('should match snapshot with correct username/password', async () => {
      const response = await getLauncherToken(username, password);

      expect(response).toMatchSnapshot();
    });

    it ('should fail with invalid username/password', () => {
      expect(getLauncherToken(username, 'invalid-password'))
        .rejects
        .toThrow();
    });
  });

  describe('exchangeToken()', () => {
    const accessToken = 'access-token';

    beforeEach(() => {
      const url = OAUTH_EXCHANGE_ENDPOINT;

      nock(url.origin)
        .matchHeader('authorization', `bearer ${accessToken}`)
        .get(url.pathname)
        .reply(200, OAUTH_EXCHANGE_SUCCESS_RESPONSE)
        .get(url.pathname)
        .reply(401, OAUTH_EXCHANGE_FAILURE_RESPONSE);
    });

    it('should match snapshot with valid access token', async () => {
      const response = await exchangeToken(accessToken);

      expect(response).toMatchSnapshot();
    });

    it('should fail with invalid authorization header', () => {
      expect(exchangeToken('invalid-access-token'))
        .rejects
        .toThrow();
    });
  });

  describe('getFortniteToken()', () => {
    const exchangeCode = 'exchange-code';

    beforeEach(() => {
      const url = OAUTH_TOKEN_ENDPOINT;
      interface Body {
        exchange_code: string;
      }

      nock(url.origin)
        .persist()
        .post(url.pathname, (body: Body) => body.exchange_code === exchangeCode)
        .reply(200, OAUTH_TOKEN_SUCCESS_RESPONSE)
        .post(url.pathname)
        .reply(400, OAUTH_TOKEN_FAILURE_RESPONSE);
    });

    it('should match snapshot with correct exchange code', async () => {
      const response = await getFortniteToken(exchangeCode);

      expect(response).toMatchSnapshot();
    });

    it('should fail with invalid exchange code', () => {
      expect(getFortniteToken('invalid-exchange-code'))
        .rejects
        .toThrow();
    });
  });
});
