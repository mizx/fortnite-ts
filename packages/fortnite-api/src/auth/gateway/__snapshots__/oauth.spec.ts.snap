// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth Gateway exchangeToken() should match snapshot with valid access token 1`] = `
Object {
  "code": "code",
  "creatingClientId": "client-id",
  "expiresInSeconds": 299,
}
`;

exports[`OAuth Gateway getFortniteToken() should match snapshot with correct exchange code 1`] = `
Object {
  "access_token": "access-token-hash",
  "account_id": "account-id-hash",
  "app": "launcher",
  "client_id": "client-id-hash",
  "client_service": "launcher",
  "expires_at": "2018-08-17T14:59:39.368Z",
  "expires_in": 28800,
  "in_app_id": "app-id-hash",
  "internal_client": true,
  "lastPasswordValidation": "2018-08-17T06:59:39.368Z",
  "refresh_expires": 1987200,
  "refresh_expires_at": "2018-09-09T06:59:39.368Z",
  "refresh_token": "refresh-token",
  "token_type": "bearer",
}
`;

exports[`OAuth Gateway getLauncherToken() should match snapshot with correct username/password 1`] = `
Object {
  "access_token": "access-token-hash",
  "account_id": "account-id-hash",
  "app": "launcher",
  "client_id": "client-id-hash",
  "client_service": "launcher",
  "expires_at": "2018-08-17T14:59:39.368Z",
  "expires_in": 28800,
  "in_app_id": "app-id-hash",
  "internal_client": true,
  "lastPasswordValidation": "2018-08-17T06:59:39.368Z",
  "refresh_expires": 1987200,
  "refresh_expires_at": "2018-09-09T06:59:39.368Z",
  "refresh_token": "refresh-token",
  "token_type": "bearer",
}
`;
